src/azul/portal_service.py

```diff
@@ -0,0 +1,73 @@
+from copy import deepcopy
+from typing import (
+    Optional,
+    Sequence,
+)
+import json
+import logging
+
+import boto3
+from botocore.exceptions import ClientError
+
+from azul import config
+from azul.types import JSON
+
+log = logging.getLogger(__name__)
+
+
+class PortalService:
+
+    def get_portal_integrations_db(self, default_db: Optional[Sequence[JSON]]) -> Sequence[JSON]:
+        """
+        Retrieve portal db from S3.
+        If no object is present at the location specified in azul.config and a default db is provided,
+        then it is uploaded and returned.
+        """
+
+        s3 = boto3.resource('s3')
+        obj = s3.Object(config.portal_integrations_db_bucket, config.portal_integrations_db_object)
+        try:
+            retrieval = obj.get()
+        except ClientError as e:
+            if e.response['Error']['Code'] == 'NoSuchKey' and default_db is not None:
+                db = default_db
+                log.info('Portal integration DB not found in S3; uploading hardcoded DB.')
+                json_bytes = bytes(json.dumps(default_db), 'UTF-8')
+                obj.put(Body=json_bytes,
+                        ContentType=f'application/json; charset=UTF-8')
+            else:
+                raise e
+        else:
+            log.info("Retrieved portal integration DB from S3")
+            json_bytes = retrieval['Body'].read()
+            db = json.loads(json_bytes.decode('UTF-8'))
+        return db
+
+    def stage_db(self, unstaged_db: Sequence[JSON]) -> Sequence[JSON]:
+        """
+        Transform portal integrations database to only contain entity_ids from the current DSS deployment stage.
+        Original is not modified.
+
+        :param unstaged_db: portal DB where the `entity_ids` fields  are dictionaries whose keys correspond to DSS
+        deployment stages.
+        :return: deep copy of that DB where the `entity_ids` fields have been replaced by the entry associated with the
+        current DSS deployment stage. If the `entity_ids` field is present but no entity ids are specified for the
+        current deployment stages, the integration is removed. Portals, however, are not removed even if they have no
+        remaining associated integrations.
+        """
+        db = deepcopy(unstaged_db)
+
+        def transform(integrations):
+            for integration in integrations:
+                if 'entity_ids' not in integration:
+                    yield integration
+                else:
+                    staged_entity_ids = integration['entity_ids'][config.dss_deployment_stage]
+                    if staged_entity_ids:
+                        integration['entity_ids'] = staged_entity_ids
+                        yield integration
+
+        for portal in db:
+            portal['integrations'] = list(transform(portal['integrations']))
```

Comment:
To match original behavior portals should only be returned by the `/integrations` endpoint if they have 1 or more integrations matching the query params `entity_type`, `integration_type`, and (optionally) `entity_ids`.

Reply:
The endpoint behavior is not an issue. Obviously, a portal with no integrations has no integrations matching the query parameters, so it would never included in the results. That remains unchanged in this PR (see [the updated integrations endpoint](https://github.com/DataBiosphere/azul/pull/1428/files#diff-a3dec4f6b35a08e7cb29086cb8766c3a)).

The question is whether portals with no integrations defined for the current deployment stage should be included in the database at all, following the deployment-specific flattening of the entity ids. Since we've established that they cannot contribute to the integrations endpoint, it would seem they should not?

Reply:
Ah, thanks for the clarification. When a new portal registers with the Portal Integrations AP, it first will post to `/portal` to register the portal details, then will post to `/portal/{portal_id}/integrations` to register the integrations, meaning before the portal's registers the first integration the database will only have the portal details with no integrations.

So yes, since portals with no integrations need to be allowed to exist in the deployment-specific database, therefore we might as well include those from the flattened default_db.